x-pack/libbeat/autodiscover/providers/aws/ec2/provider.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	awsauto "github.com/elastic/beats/x-pack/libbeat/autodiscover/providers/aws"
+	awscommon "github.com/elastic/beats/x-pack/libbeat/common/aws"
+)
+
+func init() {
+	autodiscover.Registry.AddProvider("aws_ec2", AutodiscoverBuilder)
+}
+
+// Provider implements autodiscover provider for aws EC2s.
+type Provider struct {
+	config        *awsauto.Config
+	bus           bus.Bus
+	templates     *template.Mapper
+	startListener bus.Listener
+	stopListener  bus.Listener
+	watcher       *watcher
+	uuid          uuid.UUID
+}
+
+// AutodiscoverBuilder is the main builder for this provider.
+func AutodiscoverBuilder(bus bus.Bus, uuid uuid.UUID, c *common.Config) (autodiscover.Provider, error) {
+	cfgwarn.Experimental("aws_ec2 autodiscover is experimental")
+
+	config := awsauto.DefaultConfig()
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	awsCfg, err := awscommon.GetAWSCredentials(
+		awscommon.ConfigAWS{
+			AccessKeyID:     config.AWSConfig.AccessKeyID,
+			SecretAccessKey: config.AWSConfig.SecretAccessKey,
+			SessionToken:    config.AWSConfig.SessionToken,
+			ProfileName:     config.AWSConfig.ProfileName,
+		})
+
+	// Construct MetricSet with a full regions list if there is no region specified.
+	if config.Regions == nil {
+		// set default region to make initial aws api call
+		awsCfg.Region = "us-west-1"
+		svcEC2 := ec2.New(awsCfg)
+		completeRegionsList, err := awsauto.GetRegions(svcEC2)
+		if err != nil {
+			return nil, err
+		}
+
+		config.Regions = completeRegionsList
+	}
+
+	var clients []ec2iface.ClientAPI
+	for _, region := range config.Regions {
+		if err != nil {
+			logp.Error(errors.Wrap(err, "error loading AWS config for aws_ec2 autodiscover provider"))
+		}
+		awsCfg.Region = region
+		clients = append(clients, ec2.New(awsCfg))
+	}
+
+	return internalBuilder(uuid, bus, config, newAPIFetcher(clients))
+}
+
+// internalBuilder is mainly intended for testing via mocks and stubs.
+// it can be configured to use a fetcher that doesn't actually hit the AWS API.
+func internalBuilder(uuid uuid.UUID, bus bus.Bus, config *awsauto.Config, fetcher fetcher) (*Provider, error) {
+	mapper, err := template.NewConfigMapper(config.Templates)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Provider{
+		config:    config,
+		bus:       bus,
+		templates: &mapper,
+		uuid:      uuid,
+	}
+
+	p.watcher = newWatcher(
+		fetcher,
+		config.Period,
+		p.onWatcherStart,
+		p.onWatcherStop,
+	)
+
+	return p, nil
+}
+
+// Start the autodiscover process.
+func (p *Provider) Start() {
+	p.watcher.start()
+}
+
+// Stop the autodiscover process.
+func (p *Provider) Stop() {
+	p.watcher.stop()
+}
+
+func (p *Provider) onWatcherStart(arn string, instance *ec2Instance) {
+	e := bus.Event{
+		"start":    true,
+		"provider": p.uuid,
+		"id":       arn,
+		"meta": common.MapStr{
+			"ec2":   instance.toMap(),
```

Comment:
We should probably prefix this with `aws`?

Reply:
@exekias I thought about this but `aws_elb` provider starts with `elb` hmmm Should I be consistent with aws_elb or should I be consistent with aws/ec2 metricset? 😄 Or maybe we can change `aws_elb` to be `aws.elb` as well.

Reply:
@andrewvc what do you think about this? I think is good if we have the same behavior across providers. If we use the same field names between AWS module & autodiscover that will enable us for correlation. ie Answering questions like show me the uptime status for this load balancer I'm seeing in the infra UI

Reply:
I think the AWS prefix makes sense, but we probably can't make a breaking change in 7.x for `aws_elb`.

Reply:
I just pushed a commit to add aws prefix. After this PR gets merged, I will create an issue to track the work for changing elb to aws.elb for 8.0 as a breaking change. Thanks!